### PR TITLE
processing: 3 -> 4

### PR DESCRIPTION
**Pull request by matyklug18**
###### Motivation for this change
Processing 3 is outdated, and requires Oracle JDK

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/processing/default.nix [hunk 1]
```diff
@@ -1,73 +1,94 @@
...
   pname = "processing";
-  version = "3.5.4";
+  version = "1276-4.0b1";
 
   src = fetchFromGitHub {
     owner = "processing";
-    repo = "processing";
-    rev = "processing-0270-${version}";
-    sha256 = "0cvv8jda9y8qnfcsziasyv3w7h3w22q78ihr23cm4an63ghxci58";
+    repo = "${pname}";
```
Comment (legendofmiracles) [inline pkgs/applications/graphics/processing/default.nix:39]:
```suggestion
    repo = pname;
```

File: pkgs/applications/graphics/processing/default.nix [hunk 2]
```diff
@@ -1,73 +1,94 @@
...
+  ffmpeg = fetchurl {
+    url = "https://download.processing.org/ffmpeg/4.4/ffmpeg-linux-amd64.gz";
+    name = "ffmpeg-4.4.gz";
+    sha256 = "45M5oArfpTuGONyeem9+a2bXCrfw7TEILoZ4znXrcPo=";
+  };
+in
 stdenv.mkDerivation rec {
   pname = "processing";
-  version = "3.5.4";
+  version = "1276-4.0b1";
```
Comment (legendofmiracles) [inline pkgs/applications/graphics/processing/default.nix:35]: Isn't it version 4.0?

File: pkgs/applications/graphics/processing/default.nix [hunk 3]
```diff
@@ -1,73 +1,94 @@
...
-      --replace 'Messages.showWarning' 'if (false) Messages.showWarning'
-
-    ( cd build
-      substituteInPlace build.xml --replace "jre-download," ""  # do not download jre1.8.0_144
-      mkdir -p linux/jre1.8.0_144                               # fake dir to avoid error
-      ant build )
+    tar --checkpoint=10000 -czf build/linux/jdk-11.0.12.tgz ${jdk11}
+    cp ${ant}/lib/ant/lib/{ant.jar,ant-launcher.jar} app/lib/
+    mkdir -p core/library
+    cp ${javaPackages.jogl_2_3_2}/share/java/* core/library/
```
Comment (legendofmiracles) [inline pkgs/applications/graphics/processing/default.nix:51]:
```suggestion
    ln ${javaPackages.jogl_2_3_2}/share/java/* core/library/
```

File: pkgs/applications/graphics/processing/default.nix [hunk 4]
```diff
@@ -1,73 +1,94 @@
...
+    sha256 = "W9Q5LQ0tak3cEr7fjZEtSvF/hiLwVg2+AoH1bE+e0Uw=";
+  };
+
+  batik = fetchurl {
+    name = "batik-bin-1.14.zip";
+    url = "https://archive.apache.org/dist/xmlgraphics/batik/binaries/batik-bin-1.14.zip";
+    sha256 = "vNYa4jW7QSPucGHKFWmLI2eTOnRYfmzDNzUiwmzCX28=";
+  };
 
+  ffmpeg = fetchurl {
```
Comment (legendofmiracles) [inline pkgs/applications/graphics/processing/default.nix:27]: Why can we not use the ffmpeg that nixpkgs has?

File: pkgs/applications/graphics/processing/default.nix [hunk 5]
```diff
@@ -1,73 +1,94 @@
...
+    url = "https://download.processing.org/ffmpeg/4.4/ffmpeg-linux-amd64.gz";
+    name = "ffmpeg-4.4.gz";
+    sha256 = "45M5oArfpTuGONyeem9+a2bXCrfw7TEILoZ4znXrcPo=";
+  };
```
Comment (Artturin) [inline pkgs/applications/graphics/processing/default.nix:31]: is that their own fork of ffmpeg? if its just a normal ffmpeg, add ffmpeg to arguments and copy `ffmpeg.src`

**Review by legendofmiracles — CHANGES_REQUESTED** [review]

File: pkgs/applications/graphics/processing/default.nix (hunk 1 above)
Comment (Artturin) [inline pkgs/applications/graphics/processing/default.nix:39]:
```suggestion
    repo = "processing4";
```

File: pkgs/applications/graphics/processing/default.nix [hunk 6]
```diff
@@ -1,73 +1,94 @@
...
+    cp -dpr build/linux/work $out/${pname}
     rmdir $out/${pname}/java
-    ln -s ${jdk} $out/${pname}/java
-
+    ln -s ${jdk11} $out/${pname}/java
     makeWrapper $out/${pname}/processing      $out/bin/processing \
-        --prefix XDG_DATA_DIRS : ${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name} \
-        --prefix _JAVA_OPTIONS " " -Dawt.useSystemAAFontSettings=lcd \
-        --prefix LD_LIBRARY_PATH : ${libXxf86vm}/lib
+      --prefix XDG_DATA_DIRS : ${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name} \
```
Comment (Artturin) [inline pkgs/applications/graphics/processing/default.nix:78]: use `wrapGAppsHook` and set `dontWrapGApps = true;` and add `''${gappsWrapperArgs[@]}` to here

**Comment by tomberek** [comment]
Result of `nixpkgs-review pr 145778` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>processing</li>
  </ul>
</details>

File: pkgs/applications/graphics/processing/default.nix (hunk 2 above)
Comment (matyklug18) [reply pkgs/applications/graphics/processing/default.nix:30]: I am not sure i understand. Yes, it is indeed version 4, as written in the string. The version comes from the revision in the processing/processsing4 repository, which is "1276-4.0b1". Should i change it to just `4.0b1`? Or `4.0b1-1276`? (I reuse the version string as the revision string)

Comment (legendofmiracles) [reply pkgs/applications/graphics/processing/default.nix:30]: That you reuse the version string is perfectly fine, I was referring to your commit/PR title which say it's 4 without the commit, and release candidate numbers included. But I think it's fine, was just initially confused...

**Review by legendofmiracles — COMMENTED** [review]
Builds and opens on my machine

File: pkgs/applications/graphics/processing/default.nix (hunk 6 above)
Comment (Artturin) [reply pkgs/applications/graphics/processing/default.nix:101]: https://nixos.org/manual/nixpkgs/unstable/#ssec-gnome-hooks

File: pkgs/applications/graphics/processing/default.nix [hunk 7]
```diff
@@ -1,73 +1,85 @@
...
 
   installPhase = ''
     mkdir $out
-    cp -dpR build/linux/work $out/${pname}
-
+    cp -dpr build/linux/work $out/${pname}
     rmdir $out/${pname}/java
-    ln -s ${jdk} $out/${pname}/java
-
+    ln -s ${jdk11} $out/${pname}/java
```
Comment (Artturin) [inline pkgs/applications/graphics/processing/default.nix:69]: Install all the files to `$out/share/processing`

File: pkgs/applications/graphics/processing/default.nix (hunk 2 above)
Comment (sdedovic) [reply pkgs/applications/graphics/processing/default.nix:24]: For what it's worth, I think it makes sense to change it to `4.0b1`.

If you look at the Processing 3 repo, releases are named:
- `3.5.4`
- `3.5.3`
- `3.5.2`
- etc.

but the release information and git tag also contain the revision, e.g. `3.5.4`:
- git tag: `processing-0270-3.5.4`
- release header: `PROCESSING 3.5.4 (REV 0270) - 17 January 2020`
- [link to release](https://github.com/processing/processing/releases/tag/processing-0270-3.5.4)

This is also consistent with the Processing.org download links, the version being `processing-4.0b1.zip`.

File: pkgs/applications/graphics/processing/default.nix [hunk 8]
```diff
@@ -1,73 +1,85 @@
...
+    sha256 = "cd0b82df8e7434c902ec873364bf3e9a3e6bef8b59cbf42433130d71bf1a779c";
+  };
+
+  jna = fetchurl {
+    name = "jna-5.8.0.zip";
+    url = "https://github.com/java-native-access/jna/archive/5.8.0.zip";
+    sha256 = "7cRkkxyo2Ld19Nif8oWceITt7dYMUnQlrs0R5e0AmPE=";
+  };
+
+  gluon = fetchurl {
```
Comment (sdedovic) [inline pkgs/applications/graphics/processing/default.nix:15]: Is there a reason this cannot use the nixpkgs `openjfx` packages [here](https://github.com/NixOS/nixpkgs/tree/nixos-21.05/pkgs/development/compilers/openjdk/openjfx)?

File: pkgs/applications/graphics/processing/default.nix (hunk 2 above)
Comment (sdedovic) [reply pkgs/applications/graphics/processing/default.nix:24]: Also I think we should add this as `processing4` and leave `processing` alone until Processing 4 becomes stable. And keep a `processing3` package due to many breaking changes between the two.

Comment (Artturin) [reply pkgs/applications/graphics/processing/default.nix:24]: https://processing.org/download/ lists 4.0 beta 2 as a stable release

Comment (sdedovic) [reply pkgs/applications/graphics/processing/default.nix:24]: Also, there is an alias `processing3` to `processing`, that may need to be updated if the `processing` package is for Processing 4.

**Comment by evan-goode** [comment]
Processing 4.0 beta 4 is out: https://github.com/processing/processing4/releases. They've stopped including JavaFX with it, so I don't think we need to worry about that any more, and they're now using JDK 17.

I took a stab at updating this PR for 4.0b4: https://github.com/evan-goode/nixpkgs/commit/c9ccd15360850a9c158c1176a8adefdae5c192bb.

**Comment by onny** [comment]
Please rebase

**Comment by iNecas** [comment]
In the meantime, processing 4.0.1 got released, not marked as beta anymore https://github.com/processing/processing4/releases/tag/processing-1286-4.0.1

**Comment by evan-goode** [comment]
I've updated my fork for 4.0.1, maybe I should take this over as a new PR

https://github.com/evan-goode/nixpkgs/commit/2660f2dc697e35778508a6ece9da61c5f56cf0d9

**Comment by evan-goode** [comment]
I created a new PR for Processing 4.0.1: https://github.com/NixOS/nixpkgs/pull/192148.

**Pull request closed without merging**